jdaviz/configs/specviz/plugins/unit_conversion/unit_conversion.py

```diff
@@ -0,0 +1,271 @@
+        """
+        Set up UI to have all values of currently visible spectra.
+        """
+        if len(self.dc_items) < 1:
+            self.selected_data = ""
+            self.current_flux_unit = ""
+            self.current_spectral_axis_unit = ""
+            return
+
+        self.selected_data = self.app.get_viewer("spectrum-viewer").state.reference_data.label
+
+        self.spectrum = self._viewer_data[self.selected_data]
+
+        # Set UI label to show current flux and spectral axis units.
+        self.current_flux_unit = self.spectrum.flux.unit.to_string()
+        self.current_spectral_axis_unit = self.spectrum.spectral_axis.unit.to_string()
+
+        # Populate drop down with all valid options for unit conversion.
+        self.spectral_axis_unit_equivalencies = self.create_spectral_equivalencies_list()
+        # self.flux_unit_equivalencies = self.create_flux_equivalencies_list()
+        self.flux_unit_equivalencies = self._locally_defined_flux_units()
+
+    def vue_unit_conversion(self, *args, **kwargs):
+        """
+        Runs when the `apply` button is hit. Tries to change units if `new` units are set
+        and are valid.
+        """
+
+        print(self.new_spectral_axis_unit)
+        print(self.new_flux_unit)
+        set_spectral_axis_unit = self.spectrum.spectral_axis
+        set_flux_unit = self.spectrum.flux
+
+        # Try to set new units if set and are valid.
+        if self.new_spectral_axis_unit != "" and\
+                self.new_spectral_axis_unit != self.current_spectral_axis_unit:
+            try:
+                set_spectral_axis_unit = self.spectrum.spectral_axis.to(u.Unit(self.new_spectral_axis_unit))
+            except ValueError:
+                snackbar_message = SnackbarMessage(
+                    f"Unable to convert spectral axis units for selected data. Try different units.",
+                    color="error",
+                    sender=self)
+                self.hub.broadcast(snackbar_message)
+
+                return
+        print(type(set_spectral_axis_unit))
+
+        print(self.spectrum.flux.value)
+        # Try to set new units if set and are valid.
+        if self.new_flux_unit != "" and self.new_flux_unit != self.current_flux_unit:
+            try:
+                set_flux_unit = self.spectrum.flux.to(u.Unit(self.new_flux_unit),
+                                                      equivalencies=u.spectral_density(set_spectral_axis_unit))
+            except ValueError:
+                snackbar_message = SnackbarMessage(
+                    f"Unable to convert flux units for selected data. Try different units.",
+                    color="error",
+                    sender=self)
+                self.hub.broadcast(snackbar_message)
+
+                return
+        print(set_flux_unit.value)
+        # Create new spectrum with new units.
+        converted_spec = Spectrum1D(flux=set_flux_unit, spectral_axis=set_spectral_axis_unit)
```

Comment:
I would suggest using the `_copy` method to ensure that we're maintaining any old data or e.g. redshift data on the spectrum.

```suggestion
        converted_spec = self.spectrum._copy(flux=set_flux_unit, spectral_axis=set_spectral_axis_unit)
```

Reply:
This does not transfer the `flux.unit` correctly. The `flux` values are converted but the `flux.unit` is still set to the old one.

Reply:
Never mind, I needed to set `unit=set_flux_unit.unit` as well in the `_copy` method.